### ISSUE TITLE
1d array training data fails explainer object creation

### ISSUE THREAD

**Issue by thejamesmarq** [shap/shap#487]
I have a model that is fitted on a 1d array ( e.g. shape (n,) ). When I try to create an explainer (currently tested on Kernel Explainer) with the same data that was passed to model fitting, I see an error that is raised by my preprocessor that the incoming shape of the data is incorrect.

After digging through the code, it looks like Explainer objects call match_model_to_data, using as input a modified representation of the data originally passed. This winds up taking my (n,) array and turning it into a (n,1) array. It appears the code that is doing this is here in /shap/common.py:

```python
def convert_to_data(val, keep_index=False):
    if isinstance(val, Data):
        return val
    elif type(val) == np.ndarray:
        return DenseData(val, [str(i) for i in range(val.shape[1])])
    elif str(type(val)).endswith("'pandas.core.series.Series'>"):
        return DenseData(val.values.reshape((1,len(val))), list(val.index))
    elif str(type(val)).endswith("'pandas.core.frame.DataFrame'>"):
        if keep_index:
            return DenseDataWithIndex(val.values, list(val.columns), val.index.values, val.index.name)
        else:
            return DenseData(val.values, list(val.columns))
    elif sp.sparse.issparse(val):
        if not sp.sparse.isspmatrix_csr(val):
            val = val.tocsr()
        return SparseData(val)
    else:
        assert False, "Unknown type passed as data object: "+str(type(val))
```

I would suggest that an explainer should allow a 1d array as input to prevent this issue.

**Comment by slundberg** [comment]
Could you clarify how the fitting is done? In other words what are the "samples" that are used? The background data always assumes the first dimension is the samples dimension, so a 1D array would mean only a single independent variable (feature).

**Comment by thejamesmarq** [comment]
The array is an array of documents. Embedding and padding happen as part of fit.

Actually, looking at the example here https://slundberg.github.io/shap/notebooks/deep_explainer/Keras%20LSTM%20for%20IMDB%20Sentiment%20Classification.html it looks like the expectation for Explainers might be that embedding/padding happens prior to fitting. Is that the case?

**Comment by slundberg** [comment]
That's right. It attributes importance to a set of input feature (however you define those features). If you just have a single features for each sample which is "the document" then feature attribution is not interesting. If instead you break up the document into words or word embeddings then there are many features, and so allocating importance among them is interesting.

**Comment by thejamesmarq** [comment]
Absolutely makes sense. For my use case the model does multiple tokenizations, embeddings, etc., but in the end I want to be able to assign importance to a word in the document rather to a particular derived feature from the word (for example, I want to know that the word "marital" at index 3 in a document has a particular importance rather than a particular value of an embedding output).

Also, I'm not sure how often others are doing this, but I often have multiple embeddings, tokenizations, learned threshold bucket features in my models, but when I am trying to explain a model to a layperson these details aren't as important as the presence, absence, or order of individual words.

In https://github.com/marcotcr/lime there is a LimeTextExplainer class that creates an indexed word vector and explains words by using this created vector. I'm not sure if it really makes sense to do something similar for this project.
